ruby/ql/src/experimental/decompression-api/DecompressionApi.ql

```diff
@@ -0,0 +1,51 @@
+/**
+ * @name User-controlled file decompression
+ * @description User-controlled data that flows into decompression library APIs could be dangerous
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision medium
+ * @id rb/user-controlled-file-decompression
+ * @tags security external/cwe/cwe-409
+ */
+
+import ruby
+import codeql.ruby.ApiGraphs
+import codeql.ruby.DataFlow
+import codeql.ruby.dataflow.RemoteFlowSources
+import codeql.ruby.TaintTracking
+import DataFlow::PathGraph
+
+class DecompressionApiUse extends DataFlow::Node {
+  private DataFlow::CallNode call;
+
+  // this should find the first argument of Zlib::Inflate.inflate
+  DecompressionApiUse() {
+    this = call.getArgument(0) and
+    (
+      call = API::getTopLevelMember("Zlib").getMember("Inflate").getAMethodCall("inflate") or
+      call = API::getTopLevelMember("Zip").getMember("File").getAMethodCall("open_buffer")
+    )
+  }
+
+  DataFlow::CallNode getCall() { result = call }
+}
+
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "DecompressionApiUse" }
+
+  // this predicate will be used to contstrain our query to find instances where only remote user-controlled data flows to the sink
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  // our Decompression APIs defined above will the the sinks we use for this query
+  override predicate isSink(DataFlow::Node sink) {
+    sink.(DataFlow::CallNode) instanceof DecompressionApiUse
```

Comment:
Let's not restrict the sink to be a call node. Otherwise we can only find things like `inflate(some_call())`
```suggestion
    sink instanceof DecompressionApiUse
```